storage: Fix bug where cluster gets stuck if node restarts without its data

a-robinson opened the pull request:
gossip: Don't allow two nodes to share the same address
When a new node registers with the same address as an existing node,
assume that the existing node has been replaced by the new one and
clear it out of the infoStore.

storage: Stop processing if target node doesn't contain desired store
Do so by adding a new roachpb error type that can be recognized in such
cases.

I'm open to suggestions on other approaches to the gossip commit. It looks as though our gossip depends on each piece of Info being immutable, so replacing the populated descriptor with an empty descriptor at a later TTL seemed like a better approach than trying to add a different way to remove an entry, but I may be missing a cleaner way.

I'm still working on getting a testcluster test in place. Sending this out now for feedback in the meantime. I've manually verified that this fixes the scenario in practice.

Fixes #10266

@tamird

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10544)
<!-- Reviewable:end -->

a-robinson committed:
gossip: Don't allow two nodes to share the same address

When a new node registers with the same address as an existing node,
assume that the existing node has been replaced by the new one and
clear it out of the infoStore.

a-robinson committed:
storage: Stop processing if target node doesn't contain desired store

Do so by adding a new roachpb error type that can be recognized in such
cases.

a-robinson commented:
Repushed to fix a broken test

tamird commented:
Second commit needs a test?

---

Reviewed 5 of 5 files at r1, 1 of 1 files at r2, 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[pkg/gossip/gossip.go, line 193 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rWbHEr8CHXAJS_3A:-KW4rWbHEr8CHXAJS_3B:b9sprtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L193)):*
> ```Go
> 	resolversTried map[int]struct{} // Set of attempted resolver indexes
> 	nodeDescs      map[roachpb.NodeID]*roachpb.NodeDescriptor
> 	nodeAddrs      map[util.UnresolvedAddr]roachpb.NodeID
> ```

is this a superset of `bootstrapAddrs`?

---

*[pkg/gossip/gossip.go, line 584 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tc8z4A-AVlnKUahW:-KW4tc8z4A-AVlnKUahX:bir9cpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L584)):*
> ```Go
> 	// removed from the cluster. If that's the case, remove it from our map of
> 	// nodes to prevent other parts of the system from trying to talk to it.
> 	if desc.NodeID == 0 {
> ```

`if desc == (roachpb.NodeDescriptor{})`?

---

*[pkg/gossip/gossip.go, line 605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tsYmgYw5RoEEj9dt:-KW4tsYmgYw5RoEEj9du:b-dbtu4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L605)):*
> ```Go
> 	// it from our set of tracked descriptors to ensure we don't attempt to
> 	// connect to its previous identity (as came up in issue #10266).
> 	// We ignore empty addresses for the sake of not breaking tests.
> ```

how many tests? It might be better to add a test-only node descriptor factory method.

---

*[pkg/gossip/gossip.go, line 608 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4u1znJNw1KBh9t7jt:-KW4u1znJNw1KBh9t7ju:bvdrdsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L608)):*
> ```Go
> 	if desc.Address != (util.UnresolvedAddr{}) {
> 		if oldNodeID, ok := g.nodeAddrs[desc.Address]; ok && oldNodeID != desc.NodeID {
> 			log.Infof(ctx, "removing node %d which was at same address as new node %v", oldNodeID, desc)
> ```

log the address as well?

---

*[pkg/gossip/gossip.go, line 610 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4uGknjJQPmue0JQWV:-KW4uGknjJQPmue0JQWW:bor120d) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L610)):*
> ```Go
> 			log.Infof(ctx, "removing node %d which was at same address as new node %v", oldNodeID, desc)
> 			delete(g.nodeDescs, oldNodeID)
> 			bytes, err := protoutil.Marshal(&roachpb.NodeDescriptor{})
> ```

i think you should add a comment here to explain what's going on and that this is recursive because this is a gossip callback.

---

*[pkg/gossip/gossip.go, line 620 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4uvCs014NvOBPXa6K:-KW4uvCs014NvOBPXa6L:ban66ld) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L620)):*
> ```Go
> 			}
> 		}
> 		g.nodeAddrs[desc.Address] = desc.NodeID
> ```

this map grows without bound, yeah? probably fine, just sayin'.

---

*[pkg/gossip/gossip_test.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4t1BEDFG3sb84wbq5:-KW4t1BF94Bga2MikLUS:b-mni5kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip_test.go#L70)):*
> ```Go
> 	g.SetNodeDescriptor(node1)
> 	g.SetNodeDescriptor(node2)
> 	if val, err := g.GetNodeDescriptor(1); err != nil {
> ```

shouldn't these magic numbers be `node{1,2}.NodeID`?

---

*[pkg/gossip/gossip_test.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4sq4D5VGNUDob4vBV:-KW4sq4D5VGNUDob4vBW:bn7mx9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip_test.go#L71)):*
> ```Go
> 	g.SetNodeDescriptor(node2)
> 	if val, err := g.GetNodeDescriptor(1); err != nil {
> 		t.Errorf("error fetching node 1: %s", err)
> ```

nit: all these `Errorf`s where you already have an error could just be `Error`s - the extra verbosity doesn't help and only increases the probability of rot.

---

*[pkg/gossip/gossip_test.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4t8eWBAF83nC8TW7R:-KW4t8eWBAF83nC8TW7S:bmk0q4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip_test.go#L83)):*
> ```Go
> 	// Give node3 the same address as node1, which should cause node1 to be
> 	// removed from the cluster.
> 	node3 := &roachpb.NodeDescriptor{NodeID: 3, Address: util.MakeUnresolvedAddr("tcp", "1.1.1.1:1")}
> ```

if it's the same as node1, use node1.Address rather than the same magic string.

---

*[pkg/gossip/gossip_test.go, line 92 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tGouKw_1QjRyQeA6:-KW4tGouKw_1QjRyQeA7:b-7m74ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip_test.go#L92)):*
> ```Go
> 
> 	// Quiesce the stopper now to ensure that the update has propagated before
> 	// checking whether node 1 has been removed from the infoStore.
> ```

nice!

---

*[pkg/gossip/keys.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rz8Wgv0z9pvOCkzY:-KW4rz8Wgv0z9pvOCkzZ:b-9md60a) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/keys.go#L117)):*
> ```Go
> // Returns an error if the key is not of the correct type or is not parsable.
> func NodeIDFromKey(key string) (roachpb.NodeID, error) {
> 	trimmedKey := strings.TrimPrefix(key, KeyNodeIDPrefix+separator)
> ```

oddly, two calls to TrimPrefix would be more efficient, since you wouldn't allocate the concatenated string.

---

*[pkg/gossip/keys_test.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4sFFkNouBfNA9QY8X:-KW4sFFkNouBfNA9QY8Y:b8qchzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/keys_test.go#L50)):*
> ```Go
> 		if err != nil {
> 			if tc.success {
> 				t.Errorf("%d: expected success, got error: %s", i, err)
> ```

nit: you could use subtests named after the keys instead of these indexes. that would produce better errors.

---

*[pkg/storage/store.go, line 2958 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4vTgsgurGEHcMiTxS:-KW4vTgsgurGEHcMiTxT:b-ggqz9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/21db3c85be4e0b552f255fdec7ee2380c32d962f/pkg/storage/store.go#L2958)):*
> ```Go
> }
> 
> // HandleRaftResponse handles response messages from the raft transport. It
> ```

Just remove this comment and say that it implements whatever interface it implements.

---

*[pkg/storage/store.go, line 2970 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4vbR6JU6vT5hnRCkP:-KW4vbR6JU6vT5hnRCkQ:b-kgfhcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/21db3c85be4e0b552f255fdec7ee2380c32d962f/pkg/storage/store.go#L2970)):*
> ```Go
> 			repl, err := s.GetReplica(resp.RangeID)
> 			if err != nil {
> 				return nil // not unexpected
> ```

NYC but this should really check for `roachpb.RangeNotFoundError`; anything else _is_ unexpected.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-:-KW4vmYdn6F2AsDnGNtr:bqp3fs9)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

@spencerkimball might want to take a look at the gossip changes.

---

Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/gossip/gossip.go, line 606 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9JePJKWBwyUDOHWWT:-KW9JePJKWBwyUDOHWWU:b4he422) ([raw file](https://github.com/cockroachdb/cockroach/blob/49f5dc9b74904cebea71312dc1d2f6b01e055de2/pkg/gossip/gossip.go#L606)):*
> ```Go
> 	// connect to its previous identity (as came up in issue #10266).
> 	// We ignore empty addresses for the sake of not breaking tests.
> 	if desc.Address != (util.UnresolvedAddr{}) {
> ```

Might be cleaner to add an `UnresolvedAddr.IsEmpty` helper method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-:-KW9eo9MNhGw34FT4GIg:boiuy0i)*
<!-- Sent from Reviewable.io -->

a-robinson committed:
storage: Log unexpected GetReplica errors when handling raft response

a-robinson commented:
Review status: 1 of 12 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/gossip/gossip.go, line 193 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rWbHEr8CHXAJS_3A:-KW9iuIlWeLYwtReqa4L:b9lyio6) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L193)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this a superset of `bootstrapAddrs`?
</details>
Yes, effectively. Do you think it's worth refactoring?

---

*[pkg/gossip/gossip.go, line 584 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tc8z4A-AVlnKUahW:-KW9YNsUtwzt-pQ1E04T:bnaxjxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L584)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `if desc == (roachpb.NodeDescriptor{})`?
</details>
Unfortunately roachpb.NodeDescriptor isn't comparable due to its roachpb.Attributes field

---

*[pkg/gossip/gossip.go, line 605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tsYmgYw5RoEEj9dt:-KW9nEWVwcOBxseIKTc0:b-yq09ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L605)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how many tests? It might be better to add a test-only node descriptor factory method.
</details>
It's up in the double digits in just the storage package alone. I can go through and update them all if you'd like, but it'll probably lead to wasted time in the future as more tests are added without addresses specified and the developer has to figure out why their tests are timing out.

---

*[pkg/gossip/gossip.go, line 608 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4u1znJNw1KBh9t7jt:-KW9ajX51W0VTMBQDoNw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L608)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log the address as well?
</details>
Done.

---

*[pkg/gossip/gossip.go, line 610 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4uGknjJQPmue0JQWV:-KW9b6awfSIQ1T9woOk9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L610)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think you should add a comment here to explain what's going on and that this is recursive because this is a gossip callback.
</details>
Done.

---

*[pkg/gossip/gossip.go, line 620 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4uvCs014NvOBPXa6K:-KW9ZCW4eoYx0Bh5LCKa:b-a4fgg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L620)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this map grows without bound, yeah? probably fine, just sayin'.
</details>
Yeah. I also assume it's probably fine. nodeDescs grows without bound as well, otherwise I'd remove entries from here whenever we removed them from nodeDescs

---

*[pkg/gossip/gossip.go, line 606 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9JePJKWBwyUDOHWWT:-KW9ngYV1C91aVv6c3s-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49f5dc9b74904cebea71312dc1d2f6b01e055de2/pkg/gossip/gossip.go#L606)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Might be cleaner to add an `UnresolvedAddr.IsEmpty` helper method.
</details>
Done.

---

*[pkg/gossip/gossip_test.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4t1BEDFG3sb84wbq5:-KW9bP8NhVwyJbrBhqV5:bbmeuok) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip_test.go#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> shouldn't these magic numbers be `node{1,2}.NodeID`?
</details>
I think it's pretty understandable in the context of the test, but sure

---

*[pkg/gossip/gossip_test.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4sq4D5VGNUDob4vBV:-KW9bDHcLkuD83NNkSz5:b-1eh3wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip_test.go#L71)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: all these `Errorf`s where you already have an error could just be `Error`s - the extra verbosity doesn't help and only increases the probability of rot.
</details>
Sure, done.

---

*[pkg/gossip/gossip_test.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4t8eWBAF83nC8TW7R:-KW9cA4TW6rdYmi3KB3l:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip_test.go#L83)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> if it's the same as node1, use node1.Address rather than the same magic string.
</details>
Good call, done.

---

*[pkg/gossip/keys.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rz8Wgv0z9pvOCkzY:-KW9cyunpkSrbUjkWlGF:btdek9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/keys.go#L117)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oddly, two calls to TrimPrefix would be more efficient, since you wouldn't allocate the concatenated string.
</details>
Hm, but wouldn't each call to TrimPrefix have to allocate its return value so we'd do that work twice? Also, I'd assume the compiler can concatenate two constants at compilation time.

Doing two separate trims would also mean we'd have to add a second comparison to avoid accidentally accepting keys like ":1" on which the first trim is a no-op.

---

*[pkg/gossip/keys_test.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4sFFkNouBfNA9QY8X:-KW9fNAHlMVP4gWNYaxe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/keys_test.go#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: you could use subtests named after the keys instead of these indexes. that would produce better errors.
</details>
Done.

---

*[pkg/storage/store.go, line 2958 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4vTgsgurGEHcMiTxS:-KW9owvwRmoDMHwY_E8k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21db3c85be4e0b552f255fdec7ee2380c32d962f/pkg/storage/store.go#L2958)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Just remove this comment and say that it implements whatever interface it implements.
</details>
Done.

---

*[pkg/storage/store.go, line 2970 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4vbR6JU6vT5hnRCkP:-KW9pj3cWV4nE_83heQJ:b-7jh2ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/21db3c85be4e0b552f255fdec7ee2380c32d962f/pkg/storage/store.go#L2970)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> NYC but this should really check for `roachpb.RangeNotFoundError`; anything else _is_ unexpected.
</details>
Added a log message for non-RangeNotFound errors. Look alright?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 1 of 12 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[pkg/gossip/keys.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rz8Wgv0z9pvOCkzY:-KW9qDn-JfPPOhsDedSR:b-5uzxkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/keys.go#L117)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Hm, but wouldn't each call to TrimPrefix have to allocate its return value so we'd do that work twice? Also, I'd assume the compiler can concatenate two constants at compilation time.
> 
> Doing two separate trims would also mean we'd have to add a second comparison to avoid accidentally accepting keys like ":1" on which the first trim is a no-op.
</details>
For fun, I did a quick benchmark of this. It might not be the ideal comparison, but at least in this particular case the concatenation version is better, even leaving out the extra comparison that would be needed for correctness in the trimming twice case:

```
package bench

import (
  "strings"
  "testing"
)

func BenchmarkTrimAndConcat(b *testing.B) {
  TrimAndConcat("node:33", b.N)
}

func BenchmarkTrimTwice(b *testing.B) {
  TrimTwice("node:33", b.N)
}

const KeyNodeIDPrefix = "node"
const separator = ":"

var foo int

func TrimAndConcat(key string, n int) {
  for k := 0; k < n; k++ {
    trimmedKey := strings.TrimPrefix(key, KeyNodeIDPrefix+separator)
    if trimmedKey != "" {
      foo++
    }
  }
}

func TrimTwice(key string, n int) {
  for k := 0; k < n; k++ {
    trimmedKey := strings.TrimPrefix(strings.TrimPrefix(key, KeyNodeIDPrefix), separator)
    if trimmedKey != "" {
      foo++
    }
  }
}
````

```
$ go test -bench=.
testing: warning: no tests to run
BenchmarkTrimAndConcat-8          	300000000	         4.77 ns/op
BenchmarkTrimTwice-8              	200000000	        10.2 ns/op
PASS
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r4, 7 of 7 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/gossip/gossip.go, line 193 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rWbHEr8CHXAJS_3A:-KW9rBKuSX2N_aPMKDDI:bv254jl) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L193)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Yes, effectively. Do you think it's worth refactoring?
</details>
Yeah, I think so.

---

*[pkg/gossip/gossip.go, line 584 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tc8z4A-AVlnKUahW:-KW9rQlfCshGqcKPntK2:b-2rsf0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L584)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Unfortunately roachpb.NodeDescriptor isn't comparable due to its roachpb.Attributes field
</details>
Gah. OK, can you note that in here? Basically I want to call out that this is not a node that doesn't have a node ID assigned. The comment kind of mentions it, but I still found this confusing when reading the first time through.

---

*[pkg/gossip/gossip.go, line 605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tsYmgYw5RoEEj9dt:-KW9rtc-Zq-4Gw-DHy5V:b2thm5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L605)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> It's up in the double digits in just the storage package alone. I can go through and update them all if you'd like, but it'll probably lead to wasted time in the future as more tests are added without addresses specified and the developer has to figure out why their tests are timing out.
</details>
OK. Note that the comment is now malformed "tests the many tests".

---

*[pkg/gossip/gossip.go, line 624 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9u0WQLt5QNQidi3US:-KW9u0WQLt5QNQidi3UT:bu017bt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/gossip/gossip.go#L624)):*
> ```Go
> 			} else {
> 				key := MakeNodeIDKey(oldNodeID)
> 				if err := g.addInfoLocked(key, bytes, ttlNodeDescriptorGossip); err != nil {
> ```

actually, looking at this again, introducing `SetNodeDescriptorLocked` would save substantial boilerplate here. At the same time, you could update the comment on that method that says "...gossip network and sets the infostore's node ID." - it does not set the inforstore's node ID.

---

*[pkg/gossip/gossip_test.go, line 100 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9sbsWQPee94iH48xM:-KW9sbsWQPee94iH48xN:b-veqni2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/gossip/gossip_test.go#L100)):*
> ```Go
> 	// checking whether node 1 has been removed from the infoStore.
> 	stopper.Quiesce()
> 	if val, err := g.GetNodeDescriptor(node1.NodeID); err == nil {
> ```

can you use testutils.IsError or otherwise check for the specific error you expect?

---

*[pkg/gossip/keys.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rz8Wgv0z9pvOCkzY:-KW9v5ejkfLM3-2xwWLq:b56fc9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/keys.go#L117)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> For fun, I did a quick benchmark of this. It might not be the ideal comparison, but at least in this particular case the concatenation version is better, even leaving out the extra comparison that would be needed for correctness in the trimming twice case:
> 
> ```
> package bench
> 
> import (
>   "strings"
>   "testing"
> )
> 
> func BenchmarkTrimAndConcat(b *testing.B) {
>   TrimAndConcat("node:33", b.N)
> }
> 
> func BenchmarkTrimTwice(b *testing.B) {
>   TrimTwice("node:33", b.N)
> }
> 
> const KeyNodeIDPrefix = "node"
> const separator = ":"
> 
> var foo int
> 
> func TrimAndConcat(key string, n int) {
>   for k := 0; k < n; k++ {
>     trimmedKey := strings.TrimPrefix(key, KeyNodeIDPrefix+separator)
>     if trimmedKey != "" {
>       foo++
>     }
>   }
> }
> 
> func TrimTwice(key string, n int) {
>   for k := 0; k < n; k++ {
>     trimmedKey := strings.TrimPrefix(strings.TrimPrefix(key, KeyNodeIDPrefix), separator)
>     if trimmedKey != "" {
>       foo++
>     }
>   }
> }
> ````
> 
> ```
> $ go test -bench=.
> testing: warning: no tests to run
> BenchmarkTrimAndConcat-8          	300000000	         4.77 ns/op
> BenchmarkTrimTwice-8              	200000000	        10.2 ns/op
> PASS
> ```
</details>
Cool, thanks for benchmarking. I was expecting the double-trim to be faster because of my expectation that slicing a string never allocates. Perhaps that's incorrect?

---

*[pkg/gossip/keys_test.go, line 53 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9sl4atWnw4mAfB1tw:-KW9sl4atWnw4mAfB1tx:b-c51fqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/gossip/keys_test.go#L53)):*
> ```Go
> 					t.Errorf("expected success, got error: %s", err)
> 				}
> 				return
> ```

nit: else? also, we generally try to avoid checking for "error or no error" because that can produce false negatives as the implementation evolves. instead, we typically use an `expectedErr string` to match more precisely (empty meaning nil is expected).

---

*[pkg/storage/store.go, line 2970 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4vbR6JU6vT5hnRCkP:-KW9vPokP8svc2SiCBKX:b-d74dlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/21db3c85be4e0b552f255fdec7ee2380c32d962f/pkg/storage/store.go#L2970)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Added a log message for non-RangeNotFound errors. Look alright?
</details>
Why not return the unexpected error?

---

*[pkg/util/unresolved_addr.go, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9tBpK4eRvG71af8fw:-KW9tBpK4eRvG71af8fx:bx925n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/util/unresolved_addr.go#L78)):*
> ```Go
> 
> // IsEmpty returns true if the address has no network or address specified.
> func (a *UnresolvedAddr) IsEmpty() bool {
> ```

value receiver, since you're dereferencing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 5 files at r1, 4 of 11 files at r4, 6 of 7 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/gossip/gossip.go, line 530 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWHzWIiSxIp0iubt4AS:-KWHzWIiSxIp0iubt4AT:bo7m3qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L530)):*
> ```Go
> 				return err
> 			}
> 			if desc.Address == (util.UnresolvedAddr{}) || desc.Address == g.mu.is.NodeAddr {
> ```

Use UnresolvedAddr.IsEmpty here.

---

*[pkg/gossip/gossip.go, line 605 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWHzSI2mvdt1OX6KRWK:-KWHzSI3qHIULI1pfkTd:b9zptxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L605)):*
> ```Go
> 	// it from our set of tracked descriptors to ensure we don't attempt to
> 	// connect to its previous identity (as came up in issue #10266).
> 	// We ignore empty addresses for the sake of not breaking tests the many
> ```

s/tests//

---

*[pkg/gossip/gossip.go, line 619 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWHzoxo4bvDs9Sr-GPn:-KWHzoxo4bvDs9Sr-GPo:b-yj6im4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L619)):*
> ```Go
> 			// being updated but won't lead to deadlock because it's called
> 			// asynchronously.
> 			bytes, err := protoutil.Marshal(&roachpb.NodeDescriptor{})
> ```

As long as the proto has no `required` fields, you can just use an empty byte string. 

---

*[pkg/gossip/gossip.go, line 621 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWI-F2mG0s8LlKlE8ky:-KWI-F2mG0s8LlKlE8kz:behsq6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L621)):*
> ```Go
> 			bytes, err := protoutil.Marshal(&roachpb.NodeDescriptor{})
> 			if err != nil {
> 				log.Errorf(ctx, "failed to marshal empty NodeDescriptor proto: %s", err)
> ```

This should probably be a fatal (if it's not removed completely) since any failure here should never happen, but if it did it would happen every time and we'd have reintroduced the original bug. 

---

*[pkg/gossip/keys.go, line 114 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWI0mgg-DTggiRYYOXY:-KWI0mgh9K3XdJgXLq8m:b-iuef24) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/keys.go#L114)):*
> ```Go
> }
> 
> // NodeIDFromKey attempts to extract a NodeID from the provided key.
> ```

Mention that they key should have been constructed by `MakeNodeIDKey` (and maybe move it up to be adjacent to that function). 

---

*[pkg/storage/store.go, line 2970 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4vbR6JU6vT5hnRCkP:-KWI1Ls7AVptDaTUjEoc:b2tdbni) ([raw file](https://github.com/cockroachdb/cockroach/blob/21db3c85be4e0b552f255fdec7ee2380c32d962f/pkg/storage/store.go#L2970)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Why not return the unexpected error?
</details>
Because returning an error shuts down the connection, but an unexpected error doesn't indicate anything wrong with the connection. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-:-KWI1WNbcon1zTOqaZyq:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/gossip/gossip.go, line 193 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rWbHEr8CHXAJS_3A:-KWJ4ZlR1N5EKAtGSIqE:bp4iru4) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L193)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yeah, I think so.
</details>
Ehhh, after looking at it more closely I'm a little hesitant to mess with it. Storing node IDs in bootstrapAddrs would be awkward since we wouldn't want to store the descriptors to disk along with the addrs (since that could introduce new weird cases where they change) but don't have them yet when we're regenerating bootstrapAddrs on startup. I'm going to keep them separate for simplicity unless you feel strongly about it.

---

*[pkg/gossip/gossip.go, line 584 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tc8z4A-AVlnKUahW:-KWIvBIA0gMQKJISPX6l:b-wtrmod) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L584)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Gah. OK, can you note that in here? Basically I want to call out that this is not a node that doesn't have a node ID assigned. The comment kind of mentions it, but I still found this confusing when reading the first time through.
</details>
Done. I've also added a check that the address is empty, for good measure.

---

*[pkg/gossip/gossip.go, line 605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4tsYmgYw5RoEEj9dt:-KWIvKOsRapm0sGNgJT4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L605)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK. Note that the comment is now malformed "tests the many tests".
</details>
Done.

---

*[pkg/gossip/gossip.go, line 624 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9u0WQLt5QNQidi3US:-KWIwZF1zOhn2BXRvrSz:b-673glv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/gossip/gossip.go#L624)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> actually, looking at this again, introducing `SetNodeDescriptorLocked` would save substantial boilerplate here. At the same time, you could update the comment on that method that says "...gossip network and sets the infostore's node ID." - it does not set the inforstore's node ID.
</details>
Updated the comment, but didn't add the chain of Locked methods (SetNodeDescriptorLocked and AddInfoProtoLocked to go along with addInfoLocked) because I just switched to using an empty byte array literal at Ben's suggestion.

---

*[pkg/gossip/gossip.go, line 530 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWHzWIiSxIp0iubt4AS:-KWIuwUmSehXOaPK0lfv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L530)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Use UnresolvedAddr.IsEmpty here.
</details>
Done.

---

*[pkg/gossip/gossip.go, line 605 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWHzSI2mvdt1OX6KRWK:-KWIvKqlhKNipYFlzAAA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L605)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/tests//
</details>
Done.

---

*[pkg/gossip/gossip.go, line 619 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWHzoxo4bvDs9Sr-GPn:-KWIvTZmq768FRbV65km:bwzc1x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L619)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> As long as the proto has no `required` fields, you can just use an empty byte string. 
</details>
That struck me as being a little too clever to be worth the simplification when I was initially writing this, but I've made the change now along with a comment.

---

*[pkg/gossip/gossip.go, line 621 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWI-F2mG0s8LlKlE8ky:-KWIw70KX2XSdQRvpRgG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/gossip.go#L621)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This should probably be a fatal (if it's not removed completely) since any failure here should never happen, but if it did it would happen every time and we'd have reintroduced the original bug. 
</details>
Done.

---

*[pkg/gossip/gossip_test.go, line 100 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9sbsWQPee94iH48xM:-KWIxSTStHeJWo2NMrkr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/gossip/gossip_test.go#L100)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you use testutils.IsError or otherwise check for the specific error you expect?
</details>
Done.

---

*[pkg/gossip/keys.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rz8Wgv0z9pvOCkzY:-KWIxlXHVULJFRrzX44J:b-tp77cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/keys.go#L117)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Cool, thanks for benchmarking. I was expecting the double-trim to be faster because of my expectation that slicing a string never allocates. Perhaps that's incorrect?
</details>
Well, I assume constant folding is keeping the concatenation from ever happening at runtime.

Your assumption about slicing is correct -- TrimPrefix doesn't do any allocations. So perhaps it's the function call overhead that's affecting this benchmark.

---

*[pkg/gossip/keys.go, line 114 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWI0mgg-DTggiRYYOXY:-KWIxarpFrW-AjnI7zJQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b147b020fd4d58b320ea6364f6965c1e388b4f2d/pkg/gossip/keys.go#L114)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Mention that they key should have been constructed by `MakeNodeIDKey` (and maybe move it up to be adjacent to that function). 
</details>
Done.

---

*[pkg/gossip/keys_test.go, line 53 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9sl4atWnw4mAfB1tw:-KWIzXcA-xhBA0ONv99E:bt1o7v3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/gossip/keys_test.go#L53)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: else? also, we generally try to avoid checking for "error or no error" because that can produce false negatives as the implementation evolves. instead, we typically use an `expectedErr string` to match more precisely (empty meaning nil is expected).
</details>
Switched to an else. This is a simple enough function that I'm going to skip changing up the error checking in the test this time, but I'll do that next time.

---

*[pkg/storage/store.go, line 2970 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4vbR6JU6vT5hnRCkP:-KWIzgrKz00H6H3g95Bi:b-z63a87) ([raw file](https://github.com/cockroachdb/cockroach/blob/21db3c85be4e0b552f255fdec7ee2380c32d962f/pkg/storage/store.go#L2970)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Because returning an error shuts down the connection, but an unexpected error doesn't indicate anything wrong with the connection. 
</details>
What Ben said.

---

*[pkg/util/unresolved_addr.go, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW9tBpK4eRvG71af8fw:-KWJ-HwkdU1STJtyuXz3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c65dbc2423b9ff0fb135caef5b7e06aab9bd3a9/pkg/util/unresolved_addr.go#L78)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> value receiver, since you're dereferencing.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 12 of 12 files at r7.
Review status: 5 of 12 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/gossip/gossip.go, line 193 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rWbHEr8CHXAJS_3A:-KWJGjG4CQFCgfQvVPJ_:b-kgtmen) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L193)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Ehhh, after looking at it more closely I'm a little hesitant to mess with it. Storing node IDs in bootstrapAddrs would be awkward since we wouldn't want to store the descriptors to disk along with the addrs (since that could introduce new weird cases where they change) but don't have them yet when we're regenerating bootstrapAddrs on startup. I'm going to keep them separate for simplicity unless you feel strongly about it.
</details>
I'm not really following you here. What descriptors are you talking about? AFAICT only addresses and node IDs are present here, and it seems easy enough to ignore the node IDs wherever bootstrapAddrs is currently used.

---

*[pkg/gossip/gossip.go, line 626 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWJHFlMaZTxMZLpQ02E:-KWJHFlMaZTxMZLpQ02F:buy0x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b97cde4151b0543fb65506cefb2879a712eddde0/pkg/gossip/gossip.go#L626)):*
> ```Go
> 			// asynchronously.
> 			key := MakeNodeIDKey(oldNodeID)
> 			emptyProto := []byte{}
> ```

doesn't this still allocate? I'd have expected `var emptyProto []byte`

---

*[pkg/gossip/gossip_test.go, line 103 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWJHLm3x-Vg8h07IP6U:-KWJHLm3x-Vg8h07IP6V:b9q75pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b97cde4151b0543fb65506cefb2879a712eddde0/pkg/gossip/gossip_test.go#L103)):*
> ```Go
> 	expectedErr := "unable to look up descriptor for node"
> 	if val, err := g.GetNodeDescriptor(node1.NodeID); !testutils.IsError(err, expectedErr) {
> 		t.Errorf("expected error %q fetching node %d, got node %+v", expectedErr, node1.NodeID, val)
> ```

log the unexpected error

---

*[pkg/util/unresolved_addr.go, line 79 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWJHdYktyxsFbl9aL0I:-KWJHdYktyxsFbl9aL0J:b-41ibvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b97cde4151b0543fb65506cefb2879a712eddde0/pkg/util/unresolved_addr.go#L79)):*
> ```Go
> // IsEmpty returns true if the address has no network or address specified.
> func (a UnresolvedAddr) IsEmpty() bool {
> 	return (a == (UnresolvedAddr{}))
> ```

`return a == (UnresolvedAddr{})`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->

a-robinson committed:
gossip: Refactor Gossip's nodeAddrs and bootstrapAddrs together

a-robinson commented:
Review status: 5 of 12 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/gossip/gossip.go, line 193 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KW4rWbHEr8CHXAJS_3A:-KWJkGG8EfsLf6zo_3hn:b-b07wb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/890d789d99abfd53c642e970e0bc6d88666d8d40/pkg/gossip/gossip.go#L193)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I'm not really following you here. What descriptors are you talking about? AFAICT only addresses and node IDs are present here, and it seems easy enough to ignore the node IDs wherever bootstrapAddrs is currently used.
</details>
Sorry, I meant to write "node IDs" where I wrote "descriptors".

It feels like combining the different uses makes the code a little less understandable, but I've thrown one way of refactoring it into an additional commit. Let me know what you think.

---

*[pkg/gossip/gossip.go, line 626 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWJHFlMaZTxMZLpQ02E:-KWJjivNUX4q0scV1kNo:both31z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b97cde4151b0543fb65506cefb2879a712eddde0/pkg/gossip/gossip.go#L626)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> doesn't this still allocate? I'd have expected `var emptyProto []byte`
</details>
Yeah, I just tend to like the explicitness of initializing it and the safety of not passing nil into unknown code. Switched over to avoid the allocation though, since I realize my preferences are silly.

---

*[pkg/gossip/gossip_test.go, line 103 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWJHLm3x-Vg8h07IP6U:-KWJk6jzpvx2UHbIQsOR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b97cde4151b0543fb65506cefb2879a712eddde0/pkg/gossip/gossip_test.go#L103)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log the unexpected error
</details>
Done.

---

*[pkg/util/unresolved_addr.go, line 79 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWJHdYktyxsFbl9aL0I:-KWJkAmKOogkj3AxQh2D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b97cde4151b0543fb65506cefb2879a712eddde0/pkg/util/unresolved_addr.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `return a == (UnresolvedAddr{})`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r8, 6 of 7 files at r9, 1 of 1 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/gossip/gossip.go, line 685 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWK0R7C-YJvjcOIn9Jh:-KWK0R7C-YJvjcOIn9Ji:b-9r2qsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ae760a05409838efec992d2279239baceb49ab4/pkg/gossip/gossip.go#L685)):*
> ```Go
> 			return nil, err
> 		}
> 		if nodeDescriptor.NodeID != 0 {
> ```

this is checking the empty, right? should mirror the other site that checks for empty descriptor

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-:-KWK0_Dq3SGk0C9AmaDx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/gossip/gossip.go, line 685 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544#-KWK0R7C-YJvjcOIn9Jh:-KWK2AMGuY7y7E7j26CP:b-4c5ezp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ae760a05409838efec992d2279239baceb49ab4/pkg/gossip/gossip.go#L685)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is checking the empty, right? should mirror the other site that checks for empty descriptor
</details>
Good catch, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 7 files at r13, 1 of 1 files at r14, 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10544)*
<!-- Sent from Reviewable.io -->
